extras/CatchAddTests.cmake

```diff
@@ -121,24 +120,38 @@ function(catch_discover_tests_impl)
     endforeach()
   endif()
 
-  # Parse output
-  foreach(line ${output})
-    set(test "${line}")
+  foreach(test_case ${test_cases})
+    # Remove long form brackets
+    # [[...test case list...]]
+    string(LENGTH "${test_case}" test_case_length)
+    math(EXPR test_case_length "${test_case_length} - 4")
+    string(SUBSTRING "${test_case}" 2 ${test_case_length} test_case)
```

Comment:
I was thinking more along the lines of `set` or if that doesn't work, `eval`. Something built-in to handle this.

Better yet, can't the reporter spit out a full cmake script with all the relevant `set`, `list` instructions etc.? Technically it is a security vulnerability, but I think there are better ways to attack the test script itself rather than test names

Reply:
Generating code to be `eval`'d just sounds like a horrible idea, and as you point out, it's a security issue waiting to happen and I don't think it would be wise to take this approach.

Reply:
The only reason I suggested the block syntax was to take advantage of the built-in cmake handling. If we don't plan on using it I would prefer the original list of list approach with the escaping mechanism described in this file.

Security wise I don't think it's a significant attack vector, but you can still minimize it by disallowing `]=]` syntax in the name, tag, file, etc.

Reply:
By "block syntax" I assume you mean `[=[`/`]=]`? Unfortunately, CMake is seeing those and assuming that is the content of the string and never parses further than than (which I believe is how these strings are intended to work), which I why I have to strip the brackets off in order to process the list within.

We can revert to what I had before I opened this PR (I only escaped the the `;` originally) and not use the extended bracket syntax. I admit that syntax is pretty clunky when it doesn't really work the way we would like. I left the escaping of `\`, `,`, `[`, and `]` in the CMake file since that escaping is more specific to putting it into the CTest file and may not be a necessary thing for others you may want to use the CMake reporter for a different purpose (I am assuming we should make the CMake reporter slightly more generic -- if we were to make it with the explicit purpose of using it in this CMake script then we may as well skip the CMake script entirely and just have the reporter dump the CTest script, but this is just as much of a security issue as generating CMake `set` commands)

Square brackets can't appear in tags already (I tested this, Catch2 forbids it). I'm not sure it is worth forbidding them from the test names.

It may not be a significant attack vector, but willingly introducing an attack vector just seems like a bad idea.

Reply:
> By "block syntax" I assume you mean `[=[`/`]=]`? Unfortunately, CMake is seeing those and assuming that is the content of the string and never parses further

Indeed, the block/bracket syntax is only meant to be used within a CMake script. You only see it used in generated files to be parsed or `eval`.

> I left the escaping of `\`, `,`, `[`, and `]` in the CMake file since that escaping is more specific to putting it into the CTest file

For the reporter you would have to be even more thorough unfortunately, escaping `$` and other special symbols. This escape might also be needed with the manual handling of brackets here.

> you may want to use the CMake reporter for a different purpose

The reporter should do only one thing, generate a CMake list of list. Which format is more appropriate is debatable. The choice is between the complexity of escaping list of lists and the potential security threats.

I personally think we can address the later quite easily in this case and having a cmake script would be more readable.

> if we were to make it with the explicit purpose of using it in this CMake script then we may as well skip the CMake script entirely and just have the reporter dump the CTest script

I'm ok with either. The reusability of such a script outside ctest file generation is super limited since you have to finish building the whole project at that point. The only usage of this splitting is readability, but at that point the ctest file itself is more readable.

Hooking into the ctest file generation directly could be very useful for customizing the test properties, e.g. setting `PROCESSOR` test property for individual tests. It would require expanding the macro syntax though.

Reply:
Do we just want to replace the cmake reporter with a utility (a reporter??) that just generates the ctest file then?

Reply:
I would vote for that approach. @horenmar can you chime in on this? Unfortunately I don't have a reference for other test-suites to see how they implement this, gtest for example doesn't have tight CMake support (`gtest_discover_tests` is defined in CMake).

For simplicity I would export all input variables like `${TEST_PROPERTIES}` to environment variables like `$ENV{CATCH_TEST_PROPERTIES}` instead of defining additional `--flags`.

I don't think it needs to be a separate utility, but I didn't dig deep enough in the code to know what pitfalls I am overlooking. Maybe the confusion would be that "reporter" actually outputs the output of the run, and not just the structure, but we seem to be using the JSON one in a similar way.

Reply:
@horenmar do you have an opinion on this? Would really like to get this sorted out